### DISCUSSION TITLE
Agent Codesign with Hardware Token....

### DISCUSSION THREAD

**Discussion by AttNet** [Ylianst/MeshCentral#6381] [Q&A]
Hi @ all,

I have a small issue and i hope, someone have a good idea to solve it.
I own a EV Codesign certificate. The private key is only inside a hardware token.
No chance for a private key file.
I created a small linux vm for codesigning and with a small script, i can sign files from outside.

How can i disable the authenticode codesign process from meshcentral and 
call a own shell script for the codesigning ?
like
mycodesigner.sh infile signedoutfile

How can I implement/inject this ?
Any ideas ?

Best regards
AttNet

**Comment by si458** [comment]
Sadly, the code-signing doesn't support hardware signing yet.

I've got the exact same problem next year when mine runs out!

The only thing you can do is take a copy of the agents exe from `signed-agents` folder, which is already signed with its temp code-sign certificate

Then, sign them yaself however u want, then copy them back to the `agents` folder, then restart meshcentral

Theb meshcentral will use the agents from the `agents` folder all the time

**Reply by AttNet** [reply]
Thank you, for your answer.
I know, that hardware signing is not possible.
I can do it myself, but i need an interface for that...
If you change the files, with the self signed files in signedagents folder
and you restart meshcentral. all files would be new signed from meshcentral...
meshcentras stores hashes of the files. If these not equal, the files would be created/signed new...
I was hoping there was an interface for such a case.

**Reply by si458** [reply]
Yes, that's because u are doing it wrong
U need to do exactly as I said above

Let it self code sign the exes into `signed-agents` (This will be a temp code sign and any customise u have like name, logo, etc)

Then COPY them somewhere else and codesign them with ur software and your certificate

Then copy them back, but into the `agents` folder, you might need to create this folder inside your `meshcentral-data` folder (NOT THE `signed-agents` folder!)

Then restart meshcentral.

Don't edit/update etc anything inside the `signed-agents` folder otherwise it will keep attempting to code-sign them again because they are different than what's its expecting

**Reply by AttNet** [reply]
Thank you very much for this information.
I will try it.

What is with the other applications like MeshCentralRouter.exe ?
They are in the node_modules/meshcentral/agents directory.
The are not signed in the signedagents folder.
Can i sign them and put them as signed exe back in the node_modules/meshcentral/agents folder
or also in the meshcentral-data/agents folder ?

**Reply by si458** [reply]
In theory it only code signs the windows meshagent exes

It won't code sign the meshcentralrouter.exe 

But good idea tho to get it to code sign the router exe!

Give it a try and see if it works as I'm not 100% sure, bit code sign it then put it inside the agents folder and see if u download the router from ur meshcentral is it code signed!?
